tests/cancun/eip4788_beacon_root/test_beacon_root_precompile.py

```diff
@@ -0,0 +1,243 @@
+    pre: Dict,
+    tx: Transaction,
+    post: Dict,
+):
+    """
+    Tests the beacon root precompile call using various call contexts:
+    - `CALL`
+    - `DELEGATECALL`
+    - `CALLCODE`
+    - `STATICCALL`
+    for diffent call gas amounts:
+    - exact gas (valid call)
+    - extra gas (valid call)
+    - insufficient gas (invalid call)
+
+    The expected result is that the precompile call will be executed if the gas amount is met
+    and return the correct`parent_beacon_block_root`. Otherwise the call will be invalid, and not
+    be executed. This is highlighted within storage by storing the return value of each call
+    context.
+    """
+    state_test(
+        env=env,
+        pre=pre,
+        txs=[tx],
+        post=post,
+    )
+
+
+@pytest.mark.parametrize(
+    "timestamp, valid_input",
+    [
+        (0x0C, True),  # twelve
+        (2**32, True),  # arbitrary
+        (2**64 - 2, True),  # near-max
+        (2**64 - 1, True),  # max
+        # TODO: Update t8n to un marshal > 64-bit int
+        # Exception: failed to evaluate: ERROR(10): failed un marshaling stdin
+        # (2**64, False),  # overflow
+        # Exception: failed to evaluate: ERROR(10): failed un marshaling stdin
+        # (2**64 + 1, False),  # overflow+1
+    ],
+)
+@pytest.mark.valid_from("Cancun")
+def test_beacon_root_precompile_timestamps(
+    state_test: StateTestFiller,
+    env: Environment,
+    pre: Dict,
+    tx: Transaction,
+    post: Dict,
+):
+    """
+    Tests the beacon root precompile call across for various valid and invalid timestamps.
+
+    The expected result is that the precompile call will return the correct
+    `parent_beacon_block_root` for a valid input timestamp and return the zero'd 32 bytes value
+    for an invalid input timestamp.
+    """
+    state_test(
+        env=env,
+        pre=pre,
+        txs=[tx],
+        post=post,
+    )
+
+
+@pytest.mark.parametrize(
+    "beacon_root, timestamp",
+    [
+        (12, 12),  # twelve
+        (2**32, 2**32),  # arbitrary
+        (2**64 - 2, 2**64 - 2),  # near-max
+        (2**64 - 1, 2**64 - 1),  # max
+    ],
+    indirect=["beacon_root"],
+)
+@pytest.mark.valid_from("Cancun")
+def test_beacon_root_equal_to_timestamp(
+    state_test: StateTestFiller,
+    env: Environment,
+    pre: Dict,
+    tx: Transaction,
+    post: Dict,
+):
+    """
+    Tests the beacon root precompile call where the beacon root is equal to the timestamp.
+
+    The expected result is that the precompile call will return the `parent_beacon_block_root`,
+    as all timestamps used are valid.
+    """
+    state_test(
+        env=env,
+        pre=pre,
+        txs=[tx],
+        post=post,
+    )
+
+
+@pytest.mark.parametrize(
+    "call_gas, valid_call",
+    [
+        (BEACON_ROOT_PRECOMPILE_GAS, True),
+        (BEACON_ROOT_PRECOMPILE_GAS + 1, True),
+        (BEACON_ROOT_PRECOMPILE_GAS - 1, False),
+    ],
+)
+@pytest.mark.parametrize(
+    "timestamp",
+    [
```

Comment:
Can we add a blockchain test where we almost overwrite a value in the pre-compile storage by doing:
```
Block1.timestamp = n
Block2.timestamp = n + HISTORICAL_ROOTS_MODULUS - 1
```
or
```
Block1.timestamp = n
Block2.timestamp = n + HISTORICAL_ROOTS_MODULUS + 1
```
All the checks should be done on block 2 to verify that the previous beacon root was not overwritten.

And then another one where we actually overwrite it:
```
Block1.timestamp = n
Block2.timestamp = n + HISTORICAL_ROOTS_MODULUS
```